pkg/server/apiextensions.go

```diff
@@ -245,117 +259,226 @@ func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector
 	return ret, nil
 }
 
-// Get gets a CustomResourceDefinitions in the underlying store by name. This method does not
-// support scoping to logical clusters or workspace inheritance.
-func (c *apiBindingAwareCRDLister) Get(name string) (*apiextensionsv1.CustomResourceDefinition, error) {
-	return c.crdLister.GetWithContext(context.Background(), name)
+func isPartialMetadataRequest(ctx context.Context) bool {
+	if accept := ctx.Value(acceptHeaderContextKey).(string); len(accept) > 0 {
+		if _, params, err := mime.ParseMediaType(accept); err == nil {
+			return params["as"] == "PartialObjectMetadata" || params["as"] == "PartialObjectMetadataList"
+		}
+	}
+
+	return false
 }
 
-// GetWithContext gets a CustomResourceDefinitions in the logical cluster associated with ctx by
-// name. ClusterWorkspace API inheritance is also supported: if ctx's logical cluster does not contain the
-// CRD, and if the ClusterWorkspace for ctx's logical cluster has spec.inheritFrom set, it will try to find
-// the CRD in the referenced ClusterWorkspace/logical cluster.
-func (c *apiBindingAwareCRDLister) GetWithContext(ctx context.Context, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
-	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+func (c *APIBindingAwareCRDLister) GetLatest(crd *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, error) {
+	crdKey := clusters.ToClusterAwareKey(logicalcluster.From(crd), crd.Name)
+
+	updatedCRD, err := c.crdLister.Get(crdKey)
 	if err != nil {
 		return nil, err
 	}
 
-	if strings.HasSuffix(name, ".") {
-		name = name + "core"
+	if crd.Annotations != nil {
+		if identity := crd.Annotations[apisv1alpha1.AnnotationAPIIdentityKey]; identity != "" {
```

Comment:
does this mean the identity is stored in the bound CRD? The bound CRD is only named according to the schema UID. So we might get into conflict if multiple export with different identities reference the same schema? Seems to be wrong.

Reply:
Discussed elsewhere but recording here: no, the identity annotation is not stored in etcd. This GetLatest method is more like a Refresh (happy to rename if needed). It's only currently called by apiextensions-apiserver's getOrCreateServingInfo. And it's only here to refresh from the cache to avoid a race condition. The flow is:

1. CustomResourceHandler gets the CRD (using our lister)
2. CustomResourceHandler wants to get/create serving info for the given *CustomResourceDefinition
3. CustomResourceHandler wants to refresh from the lister (cache) to make sure it's not operating on stale data (because another goroutine maybe updated it in between step 1 and now)

Because the CRD came from our lister, it went through all the logic for identity, partial metadata, APIExports, etc. When refreshing from the cache, we need to make sure the refreshed CRD has the identity annotation from the original, and we trim the schema if it was originally for a partial metadata request.